theano/tensor/opt_uncanonicalize.py

```diff
@@ -98,3 +99,25 @@ def local_max_to_min(node):
                                  max.owner.op.axis)(neg.owner.inputs[0])]
 
     return False
+
+
+@register_uncanonicalize
+@gof.local_optimizer([T.alloc])
+def local_alloc_dimshuffle(node):
+    if node.op == T.alloc:
+        input_ = node.inputs[0]
+        if isinstance(input_.owner.op, DimShuffle):
+            # check if it only adds dimension to the left
+            pattern = input_.type.broadcastable
+            if not pattern[0] :
```

Comment:
The way to check the change that the dimshuffle do is by doing input_.owner.op.new_order. Go see the docstring of the class dimshuffle in the file tensor/elemwise.py to know how it work.